simulator/performance_manager.go

```diff
@@ -17,19 +17,247 @@ limitations under the License.
 package simulator
 
 import (
+	"math/rand"
+	"strconv"
+	"time"
+
 	"github.com/vmware/govmomi/object"
 	"github.com/vmware/govmomi/simulator/esx"
+	"github.com/vmware/govmomi/simulator/vpx"
+	"github.com/vmware/govmomi/vim25/methods"
 	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/soap"
 	"github.com/vmware/govmomi/vim25/types"
 )
 
+var realtimeProviderSummary = types.PerfProviderSummary{
+	CurrentSupported: true,
+	SummarySupported: true,
+	RefreshRate:      20,
+}
+
+var historicProviderSummary = types.PerfProviderSummary{
+	CurrentSupported: false,
+	SummarySupported: true,
+	RefreshRate:      -1,
+}
+
 type PerformanceManager struct {
 	mo.PerformanceManager
+	vmMetrics         []types.PerfMetricId
+	hostMetrics       []types.PerfMetricId
+	rpMetrics         []types.PerfMetricId
+	clusterMetrics    []types.PerfMetricId
+	datastoreMetrics  []types.PerfMetricId
+	datacenterMetrics []types.PerfMetricId
+	perfCounterIndex  map[int32]types.PerfCounterInfo
+	metricData        map[string]map[int32][]int64
 }
 
 func NewPerformanceManager(ref types.ManagedObjectReference) object.Reference {
 	m := &PerformanceManager{}
 	m.Self = ref
-	m.PerfCounter = esx.PerfCounter
+	if Map.IsESX() {
+		m.PerfCounter = esx.PerfCounter[:]
+		m.hostMetrics = esx.HostMetrics[:]
+		m.vmMetrics = esx.VmMetrics[:]
+		m.rpMetrics = esx.ResourcePoolMetrics[:]
+		m.metricData = esx.MetricData
+	} else {
+		m.PerfCounter = vpx.PerfCounter[:]
+		m.hostMetrics = vpx.HostMetrics[:]
+		m.vmMetrics = vpx.VmMetrics[:]
+		m.rpMetrics = vpx.ResourcePoolMetrics[:]
+		m.clusterMetrics = vpx.ClusterMetrics[:]
+		m.datastoreMetrics = vpx.DatastoreMetrics[:]
+		m.datacenterMetrics = vpx.DatacenterMetrics[:]
+		m.metricData = vpx.MetricData
+	}
+	m.perfCounterIndex = make(map[int32]types.PerfCounterInfo, len(m.PerfCounter))
+	for _, p := range m.PerfCounter {
+		m.perfCounterIndex[p.Key] = p
+	}
 	return m
 }
+
+func (p *PerformanceManager) QueryPerfCounter(ctx *Context, req *types.QueryPerfCounter) soap.HasFault {
+	body := new(methods.QueryPerfCounterBody)
+	body.Req = req
+	body.Res.Returnval = make([]types.PerfCounterInfo, len(req.CounterId))
+	for i, id := range req.CounterId {
+		if info, ok := p.perfCounterIndex[id]; !ok {
+			body.Fault_ = Fault("", &types.InvalidArgument{
+				InvalidProperty: "CounterId",
+			})
+			return body
+		} else {
+			body.Res.Returnval[i] = info
+		}
+	}
+	return body
+}
+
+func (p *PerformanceManager) QueryPerfProviderSummary(ctx *Context, req *types.QueryPerfProviderSummary) soap.HasFault {
+	body := new(methods.QueryPerfProviderSummaryBody)
+	body.Req = req
+	body.Res = new(types.QueryPerfProviderSummaryResponse)
+
+	// The entity must exist
+	if Map.Get(req.Entity) == nil {
+		body.Fault_ = Fault("", &types.InvalidArgument{
+			InvalidProperty: "Entity",
+		})
+		return body
+	}
+
+	switch req.Entity.Type {
+	case "VirtualMachine":
```

Comment:
Looks like instead of fallthrough, this could be:
```go
case "VirtualMachine", "HostSystem", "ResourcePool":
    body.Res.Returnval = realtimeProviderSummary
```
Functionally the same, but personally try to avoid use of `fallthrough`.